### PR TITLE
WINC-860: [hack] Automate pre-release version updates

### PR DESCRIPTION
**Pull request by wgahnagl**
There are many places in a number of files where the versions of the WMCO and OCP need to be updated for release. These should be automated so it no longer has to be done manually.

### CODE REVIEW

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:1]: What shell is this targeted at?

Please ensure that the script will exit on the first error.

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:6]: It is unclear how to use this. Do you literally mean `x`, `y` or `z`? Is there an assumption that you will be in a `release-4.x` branch when executing this?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @wgahnagl. Please address my comments.

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:27]: Please add a comment explaining what these statements are doing.

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:52]: Is there a jq equivalent that you can use?

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:100]: Are you providing an input if the user does not specify `x`, `y` or `z`? Do we really need this sort of nagging rather than just erroring out?

**Review by aravindhp — CHANGES_REQUESTED** [review]
@wgahnagl please add a comment with examples in places where you are using regexes. I will defer to @saifshaikh48 @alinaryan on the release steps.

File: hack/pre-release.sh
Comment (aravindhp) [reply hack/pre-release.sh:69]: This has not been answered.

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:22]: Please add a comment about why you are setting this to 1, 2 or 3.

File: hack/pre-release.sh
Comment (alinaryan) [reply hack/pre-release.sh:100]:
```suggestion
  if [ -z $1 ]; then
  echo "Stream version not set"
  exit 1
  fi
```

**Review by alinaryan — COMMENTED** [review]
Thanks for working on this - See my suggestions

File: hack/pre-release.sh
Comment (saifshaikh48) [reply hack/pre-release.sh:1]: The skipRange still needs to be handled

File: hack/pre-release.sh
Comment (saifshaikh48) [inline hack/pre-release.sh:107]: I get an error when I run the script as is
```
$ ./hack/pre-release.sh x
./hack/pre-release.sh: line 107: STREAM: unbound variable
```

**Review by saifshaikh48 — COMMENTED** [review]
Should we make the script smarter instead of incrementing anything +1? For example, if i run this on master, which has a mix of OCP 4.13 and 4.14 sprinkled throughout the codebase, it bumps them to 4.14 and 4.15 respectively. IMO it should consistently to bring the version value to a particular version

File: hack/pre-release.sh
Comment (aravindhp) [reply hack/pre-release.sh:70]: This continues to not be answered.

File: hack/pre-release.sh
Comment (wgahnagl) [reply hack/pre-release.sh:108]: I get the SkipRange here
https://github.com/openshift/windows-machine-config-operator/blob/195b873a7ebeb69670adbb4290cb7b0e914ad4fd/hack/pre-release.sh#L32 

and update it here 
https://github.com/openshift/windows-machine-config-operator/blob/195b873a7ebeb69670adbb4290cb7b0e914ad4fd/hack/pre-release.sh#L45 
Is there somewhere else I'm missing, or should I add a comment to make it more clear ? 

I was thinking of keeping it to that one version because ideally the versions wouldn't be mixed up like that, and it would be difficult to write regexes for each individual occurrence of each version 
If master has to have that mix of versions though, I could figure it out!

Comment (saifshaikh48) [reply hack/pre-release.sh:108]: When I run `./hack/pre-release.sh x` (after fixing the unbounded variable), the skip version is not being edited. I suspect this is because you are making the changes directly to the bundle CSV in `update_WMCO_version` - when you run `make bundle` at the end of the script it will overwrite such changes because they are not being done to the base CSV

Comment (saifshaikh48) [reply hack/pre-release.sh:108]: Also shouldn't the lower bound of the skipRange also be updated?

Comment (wgahnagl) [reply hack/pre-release.sh:108]: https://github.com/openshift/windows-machine-config-operator/blob/195b873a7ebeb69670adbb4290cb7b0e914ad4fd/hack/pre-release.sh#L42 
this line updates the upper bound to the new version

and this line https://github.com/openshift/windows-machine-config-operator/blob/195b873a7ebeb69670adbb4290cb7b0e914ad4fd/hack/pre-release.sh#L45 
updates the lower bound to the current version 

I can add a comment on this to make it make more sense

File: hack/pre-release.sh
Comment (wgahnagl) [reply hack/pre-release.sh:71]: I'm getting this from a comment inside the clusterserviceversion.yaml (the description [here](https://github.com/openshift/windows-machine-config-operator/blob/cc553e29d3bc16af106392e013e77c3ac4c5428d/config/manifests/bases/windows-machine-config-operator.clusterserviceversion.yaml#L37)) 
If I were to use JQ I would end up searching the comment anyway. 
Where I'm getting this information is arbitrary though, and I could get it from somewhere else, but I figured that getting it from the doc comment would be a reliable source.

File: hack/pre-release.sh
Comment (alinaryan) [reply hack/pre-release.sh:109]: Adding here what we talked about in Slack DMs:

bundle csv = bundle/manifests/windows-machine-config-operator.clusterserviceversion.yaml
base csv = config/manifests/bases/windows-machine-config-operator.clusterserviceversion.yaml

You can't directly edit the bundle csv. Changes need to be made in the base csv and then they will be applied to the bundle csv with the make bundle command. It's similar to CI release repo - the jobs need to be edited by editing the config and then you run make jobs to get the job file added

File: hack/pre-release.sh
Comment (aravindhp) [reply hack/pre-release.sh:71]: I was not asking you to use `jq` but asking if there is a similar utility for processing yaml files. Given the amount of search and replacing with regexes you are doing here, I am wondering if there is a more elegant way of making these changes. Saying you need an utility as a pre-req is acceptable given it will make the code easier to read, understand and maintain. The way it stands, it looks pretty fragile and prone to breakage.

PS: This is a general comment and just specific to this line.

Moreover getting the version from a comment is not advisable. What if we decide to change that comment. Lets figure out a better spot to get the version from. It needn't be from the CSV but elsewhere.

Comment (aravindhp) [reply hack/pre-release.sh:71]: Please add a comment saying what exactly you are looking for with your search statements. It is in no way obvious that you were picking the version from `* OCP 4.14 cluster running on Azure, AWS, GCP or vSphere configured with hybrid OVN Kubernetes networking`. Do this for all the lines.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: hack/pre-release.sh
Comment (wgahnagl) [reply hack/pre-release.sh:73]: how about from the gitmodules [here](https://github.com/openshift/windows-machine-config-operator/blob/cc553e29d3bc16af106392e013e77c3ac4c5428d/.gitmodules#L4)

File: hack/update_submodules.sh
Comment (aravindhp) [inline hack/update_submodules.sh:129]: Nice work, @wgahnagl. Just some early feedback... this sort of refactor should be done in a separate commit.

File: hack/pre-release.sh
Comment (alinaryan) [inline hack/pre-release.sh:71]: nit: extra newline

**Review by alinaryan — COMMENTED** [review]
LGTM, just add comments for the functions you added in common.sh and make sure the commit message is capitalized

File: hack/common.sh
Comment (saifshaikh48) [inline hack/common.sh:48]: create_git_branch --> create_git_branch_list

File: hack/pre-release.sh
Comment (saifshaikh48) [inline hack/pre-release.sh:12]: nit: indent level

File: hack/pre-release.sh
Comment (saifshaikh48) [inline hack/pre-release.sh:19]: This can be made clearer. The columns mean what? I assume major, minor, and patch version

File: hack/pre-release.sh
Comment (saifshaikh48) [inline hack/pre-release.sh:8]: you can move the usage() func to the top of the file and get remove this comment block so we don't have to keep 2 places updated

File: hack/update_submodules.sh
Comment (aravindhp) [inline hack/update_submodules.sh:131]: If the function returns 1 does it automatically exit here? From what I understand, return exits terminates the subshell. Don't you have to do some return value checking here?

File: hack/common.sh
Comment (aravindhp) [inline hack/common.sh:36]: Please add a space after `()`. Applies to all the function where the space is not present.

Please add a comment for the function.

File: hack/common.sh
Comment (aravindhp) [inline hack/common.sh:46]: I don't follow what this means. Do you mean a branch name prefix?

File: hack/common.sh
Comment (aravindhp) [inline hack/common.sh:48]: The usage does not match your function description.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Left a review for the refactor.

File: hack/pre-release.sh
Comment (saifshaikh48) [inline hack/pre-release.sh:129]: shouldn't there be a separate subcommand that bumps these?

File: hack/pre-release.sh
Comment (aravindhp) [reply hack/pre-release.sh:132]: Do we really need to do recursion here?

File: hack/update_submodules.sh
Comment (wgahnagl) [reply hack/update_submodules.sh:131]: because we set the -e at the top, the return exits the subshell, and then since a command exited with a non-zero, the -e exits the program.

File: hack/common.sh
Comment (wgahnagl) [reply hack/common.sh:48]: I'm not sure what you mean by this :thinking:

Comment (aravindhp) [reply hack/common.sh:48]: Just that the function description and usage together is confusing. What do you mean by `the versions to create` when you specify `release-4.10`? Isn't the `release-4.10` branch already created? From what I can tell `release-4.10` is the branch from which you want to create another branch with a set of commits. That is not clear from your description.

File: hack/pre-release.sh
Comment (saifshaikh48) [inline hack/pre-release.sh:45]: nit: indentation

File: hack/pre-release.sh
Comment (saifshaikh48) [inline hack/pre-release.sh:133]: comment needs updating with baseimage flag

**Review by saifshaikh48 — COMMENTED** [review]
Almost LGTM, just a few minor comments. I think the arg parsing can be reworked to be more intuitive (you can take inspiration from some of our other hack scripts) but it's not a blocker for me. Nice work overall

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:80]: Same comment about `$stream_var`

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:112]: Nit: I suggest making this into a helper function and have a for loop here going over the set of files.

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:150]: What is going on here? 😅

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:18]: Do we really need this level of complexity? Why would we need to multiple releases at the same time. It is fine if we just handle `./pre-release.sh <branch> <stream>` i.e. do the updates independently. I think this will simplify some of the code below.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:74]: Why are you overwriting the global `WMCO_VERSION` variable here? This should just be a local variable.

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:154]: Why does this have to be divisible by 2?

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:72]: Please add a function doc.

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:73]: Please check for inputs

File: hack/pre-release.sh
Comment (aravindhp) [inline hack/pre-release.sh:27]: Please add a function doc

**Review by aravindhp — CHANGES_REQUESTED** [review]

**Review by aravindhp — APPROVED** [review]
Nice work, @wgahnagl

**Comment by aravindhp** [comment]
/override ci/prow/aws-e2e-upgrade ci/prow/azure-e2e-ccm-install 

The changes in the PR does not affect CI or the  operator

**Comment by aravindhp** [comment]
/hold

For lgtm from @saifshaikh48

File: hack/pre-release.sh
Comment (saifshaikh48) [inline hack/pre-release.sh:13]: You’d want to move this to where you moved the variable setting

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

1 comment but I won’t hold up the PR for it

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

1 comment but I won’t hold up the PR for it

File: hack/pre-release.sh
Comment (aravindhp) [reply hack/pre-release.sh]: FTR, we should hold up PRs for something that was overlooked and easily fixed in the current PR itself. Opening a follow up a PR to fix moving of comment is unnecessary overhead for everyone.

**Review by aravindhp — APPROVED** [review]
/lgtm

**Pull request merged**
